data/profile-physlr.Rmd

```diff
@@ -0,0 +1,384 @@
+new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
+if(length(new.packages)) install.packages(new.packages)
+
+library("dplyr")
+library("ggplot2")
+library("stringr")
+
+knit_print.data.frame <- function(x, ...) kable(x) %>% paste(collapse = "\n") %>% asis_output
+path <- params$path
+
+```
+
+## General information:
+
+Profiling of Physlr steps, based on reports provided by time command stored in *.time files in the input directory.
+
+```{r echo=FALSE, message=FALSE, warning=FALSE}
+
+m <- matrix(ncol = 7)
+df = data.frame(m)
+colnames(df)=c("step",
+               "total_time_min",
+               "time_hr",
+               "time_min",
+               "time_sec",
+               "mem",
+               "pl")
+
+#path <- "/projects/btl/jowong/github/physlr/data_hg004_stlfr/v2release"
+
+if (substr(path,nchar(path),nchar(path)) != "/")
+  path <- paste(path,"/",sep="")
+ldir <-  normalizePath(path)
+
+
+```
+
+The profiling directory including *.time files:
+
+```{r echo=FALSE}
+cat(ldir)
+
+finf <- file.info(dir(path = ldir, full.names = TRUE), extra_cols = FALSE)
+sorted_finf <- finf[with(finf, order(as.POSIXct(ctime))), ]
+sorted_files <- rownames(sorted_finf)
+sorted_time_files <- sorted_files[substr(sorted_files,
+                                         nchar(sorted_files)-4,nchar(sorted_files)) == ".time"]
+
+# Or alternatively:
+#file.info("/projects/btl/jowong/github/physlr/data_hg004_stlfr/v2release/hg004_k40.histogram.time")
+#details = file.info(list.files(path, pattern="*.time", full.names = TRUE))
+#details = details[with(details, order(as.POSIXct(ctime))), ]
+#files = rownames(details)
+
+```
+
+Unacceptable *.time files if any:
+```{r echo=FALSE}
+good = TRUE
+
+for (i in 1:length(sorted_time_files))
+{
+  lines_of_file <- readLines(sorted_time_files[i])
+  if (length(lines_of_file) != 23)
+  {
+    cat(paste("The time-file is corrupted: \n", sorted_time_files[i]))
+    next
+  }
+  ## Exit status
+  exit_line_no = grep('Exit', lines_of_file)
+  exit_status <- substr(lines_of_file[exit_line_no],
+                        nchar(lines_of_file[exit_line_no]),
+                        nchar(lines_of_file[exit_line_no]))
+  if (exit_status != "0")
+  {
+    cat("The job you are profiling did not exit normally, relative file:\n")
+    print(sorted_time_files[i])
+    next
+  }
+
+  ## Time
+  time_line_no = grep('Elapsed', lines_of_file)
+  time_line <- lines_of_file[time_line_no]
+  time = str_match(
+    time_line,
+    regex(
+      '(?<=or m:ss[[:punct:]][[:punct:]]\ ).*',
+      ignore.case = TRUE,
+      perl = TRUE
+    )
+  )[1, 1]
+
+  len = nchar(time_line)
+  colon_marks = str_match_all(time, regex(':',
+                                          ignore.case = TRUE, perl = TRUE))[[1]]
+
+  if (dim(colon_marks)[1] == 0)
+  {
+    cat(paste("Unexpected format"))
+  } else if (dim(colon_marks)[1] == 2)
+  {
+    df[i, "time_sec"] = strtoi(substr(time, nchar(time) - 1, nchar(time)), base =
```

Comment:
Couldn't you both check which format the time string conforms to and extract the numbers in one go?

```
> n
[1] "10:40:22"
> str_match(n, regex('([:digit:]+)\\:([:digit:]+)\\:([:digit:]+)'))
     [,1]       [,2] [,3] [,4]
[1,] "10:40:22" "10" "40" "22"
```

Reply:
the two formats are different... How do I detect the format and extract the numbers in one go?
format 1: `HH:MM:SS`
format 2: `MM:SS.MS` which has only one `:` and also a `.`

Reply:
You just need two regex expressions with the capture groups - that will tell you which format the string conforms to. After you do the `str_match`, you can check each for being non-NA with if-statements. Then, once you've entered the appropriate condition, you have easy access to the numbers you need.
Roughly what I was thinking:
```
hms <- str_match(...)
msms <- str_match(...)
if (is.na(hms[1,1)) {
...Parse as h,m,s..
} else if (is.na(msms[1,1]) {
...Parse as mmssms...
}
```

Reply:
I though you mean with a single call to `str_match()`. I just changed it to something like this.